fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -877,64 +878,55 @@ ACTOR Future<Void> preresolutionProcessing(CommitBatchContext* self) {
+std::pair<EncryptCipherDomainName, EncryptCipherDomainId> getEncryptDetailsFromMutationRef(ProxyCommitData* commitData,
+                                                                                           MutationRef m) {
+	std::pair<EncryptCipherDomainName, EncryptCipherDomainId> details(EncryptCipherDomainName(),
+	                                                                  INVALID_ENCRYPT_DOMAIN_ID);
 
 	// Possible scenarios:
 	// 1. Encryption domain (Tenant details) weren't explicitly provided, extract Tenant details using
 	// TenantPrefix (first 8 bytes of FDBKey)
-	// 2. Encryption domain isn't available, leverage 'default encrypiton domain'
+	// 2. Encryption domain isn't available, leverage 'default encryption domain'
+
 	bool useDefaultEncryptionDomain = true;
 	if (isSingleKeyMutation((MutationRef::Type)m.type)) {
+		ASSERT_NE((MutationRef::Type)m.type, MutationRef::Type::ClearRange);
+
 		if (isSystemKey(m.param1)) {
 			// Encryption domain == FDB SystemKeyspace encryption domain
-			details.first = EncryptCipherDomainName(FDB_DEFAULT_ENCRYPT_DOMAIN_NAME);
+			details.first = EncryptCipherDomainName(FDB_SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_NAME_REF);
 			details.second = SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID;
-		} else if ((MutationRef::Type)m.type == MutationRef::Type::ClearRange) {
-			// FIXME: Handle Clear-range transaction, actions needed:
-			// 1. Transaction range can spawn multiple encryption domains (tenants)
-			// 2. Transaction can be a multi-key transaction spawning multiple tenants
-			// For now fallback to 'default encryption domain'
-			ASSERT(useDefaultEncryptionDomain);
-
-			CODE_PROBE(true, "SingleKey ClearRange mutation encryption");
 		} else if (m.param1.size() >= TENANT_PREFIX_SIZE) {
 			// Parse mutation key to determine mutation encryption domain
 			StringRef prefix = m.param1.substr(0, TENANT_PREFIX_SIZE);
-			int64_t tenantId = TenantMapEntry::prefixToId(prefix, false /* enforceValidTenantId */);
-
+			int64_t tenantId = TenantMapEntry::prefixToId(prefix, EnforceValidTenantId::False);
 			if (tenantId != TenantInfo::INVALID_TENANT) {
-				ErrorOr<Optional<std::pair<TenantName, int64_t>>> result =
-				    getTenantDetails(commitData, tenantId, false);
-				if (!result.isError() && result.get().present()) {
-					details.first = result.get().get().first;
-					details.second = result.get().get().second;
+				Optional<TenantName> tenantName = getTenantName(commitData, tenantId);
+				if (tenantName.present()) {
+					details.first = tenantName.get();
+					details.second = tenantId;
+
 					useDefaultEncryptionDomain = false;
 				}
 			} else if (commitData->db->get().client.tenantMode == TenantMode::REQUIRED) {
-				// Disallow non-tenant non-clearrange transaction if cluster is configured with 'TenantMode::Required'
+				// Disallow non-tenant non-ClearRange transaction, when cluster is configured with
+				// 'TenantMode::Required'
```

Comment:
I don't think we want this block. It should still be possible to run operations outside of tenants with required mode.

Reply:
If I recollect offline conversation, my understanding is that `required` mode wouldn't be allowed to commit such transactions, unless it is a ClearRange? If mistaken, can enable this code block.

Reply:
The proposal in the raw access document does not require any knowledge of the tenant mode. Instead, we only use the state about which tenant was specified and/or what tenants are present in the cluster to make our decision.

In this case, we would let the operation proceed as normal and encrypt it as if it were in the default region.

Reply:
I can update the code if it is necessary, however, from cluster security aspect, `tenant_required` mode `should` discourage non-ClearRange transactions that aren't associated with any tenant. 

But, if we think it is a valid usecase, will allow it. 

Thanks

Reply:
From our last discussion on Thursday, I had mentioned some of the use cases for raw access even in required mode. I don't think there is a particular security issue that is concerning (only trusted clients could perform raw access mutations, and they can already do fairly arbitrary things), but if it were necessary we could protect against an accidental misuse risk by preventing raw access to all normal keys by default and requiring a knob to enable it.

Reply:
Sounds good. 
I'm hoping you or someone from the team is tracking work to implement measure to have better control over raw access to normal keys, even from trusted clients. Also, as we discussed in past, there is a desire to restrict or eliminate need for trusted clients in future, hence, may be for long term we think of alternate ways to achieve the goal possibly without raw-access to normal keys for `tenant_required` mode cluster at the least.

Reply:
I'm not sure what other work needs to happen relative to raw access and trusted clients, but if some requirement comes up I'm happy to be a part of that. I don't think we need to worry about the elimination of trusted clients, as that probably shouldn't interfere with any of the use-cases we discussed for raw access.